utils/zero_wipe.go

```diff
@@ -53,45 +52,56 @@ func (z *ZeroWipe) Wipe(ctx context.Context, logicalName string) error {
 
-	var bytesWritten int64
+	var totalBytesWritten int64
+	var bytesSinceLastPrint int64
 	start := time.Now()
 
-	// rewind cassete tape ;=)
-	file.Seek(0, 0)
+	// rewind cassette tape ;=)
+	whence = 0 // 0 means relative to the origin of the file, check Seek function
+	_, err = file.Seek(0, whence)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
 
-	// Writting zeroes loop
-	i := 0
-	for bytesWritten < partitionSize {
-		n, err := file.Write(buffer)
+	// Writing zeroes loop
+	bytesSinceLastPrint = 0
+	for totalBytesWritten < partitionSize {
+		bytesWritten, err := file.Write(buffer)
 		if err != nil {
-			if strings.Contains(err.Error(), "no space left on device") { //syscall.ENOSPC
-				// If partitionSize % bufferSize is not 0 the last write loop will not write all the buffer size
-				log.Println("we have reached the end of the device.")
+			if strings.Contains(err.Error(), "no space left on device") { // syscall.ENOSPC
+				// Check if the last write loop didn't write all the buffer size, the reason is partitionSize % bufferSize is not 0
+				if totalBytesWritten+int64(bytesWritten) == partitionSize {
+					log.Println("we have reached the end of the devices.")
+					return nil
+				} else {
+					log.Println("we have a problem with the write to disk: ", err)
+				}
 			} else {
 				// Other errors
-				log.Println("failed to write to disk:", err)
+				log.Println("failed to write to disk: ", err)
 			}
 			return err
 		}
-		bytesWritten += int64(n)
-		i++
-		// Calculate progress and ETA
-		progress := float64(bytesWritten) / float64(partitionSize) * 100
-		elapsed := time.Since(start).Seconds()
-		speed := float64(n*i) / elapsed                                              // Speed in bytes per second
-		remainingSeconds := (float64(partitionSize) - float64(bytesWritten)) / speed // Remaining time in seconds
-		remainingHours := float64(remainingSeconds / 3600)
-		mbPerSecond := speed / (1024 * 1024)
+		totalBytesWritten += int64(bytesWritten)
+		bytesSinceLastPrint += int64(bytesWritten)
 
 		// Print progress each 10 seconds
 		if time.Since(start) >= 10*time.Second {
+			// Calculate progress and ETA
+			progress := float64(totalBytesWritten) / float64(partitionSize) * TotalPercentage
+			elapsed := time.Since(start).Seconds()
+			speed := float64(bytesSinceLastPrint) / elapsed                                   // Speed in bytes per second
+			remainingSeconds := (float64(partitionSize) - float64(totalBytesWritten)) / speed // Remaining time in seconds
+			remainingHours := float64(remainingSeconds / SecondsPerHour)
+			mbPerSecond := speed / OneMBinBytes
 			log.Printf("%s | Progress: %.2f%% | Speed: %.2f MB/s | Estimated time left: %.2f hour(s)\n", partitionPath, progress, mbPerSecond, remainingHours)
 			start = time.Now()
-			i = 0
+			bytesSinceLastPrint = 0
 		}
 	}
 
-	log.Println("Device has been completely overwritten with zeros")
+	log.Printf("%s has been completely overwritten with zeros", partitionPath)
```

Comment:
Can we pass in the logger from the top level, so these logs will be structured and formatted?

Reply:
I'm not clear on the use of log vs. ftmt in ironlib. For example, here https://github.com/metal-toolbox/ironlib/blob/main/utils/lshw.go#L365 and https://github.com/metal-toolbox/ironlib/blob/main/utils/mlxup.go#L82 use log, and here fmt:

```
device.go: fmt.Printf("util: %s, path: %s %s[err]%s - %s\n", dmiName, dmiPath, red, reset, dmiErr.Error())
device.go: fmt.Printf("util: %s, path: %s %s[ok]%s\n", dmiName, dmiPath, green, reset)
device.go: fmt.Printf("util: %s, path: %s %s[err]%s - %s\n", name, uPath, red, reset, err.Error())
device.go: fmt.Printf("util: %s, path: %s %s[ok]%s\n", name, uPath, green, reset)
firmware/bios_checksum.go: fmt.Printf("dir-walk: %s\n", path)
```

 Is there a better way?

Reply:
I guess this can be done in separate sweep to standardise using the passed in logger, ignore my comment for now.